File: facade/src/main/java/org/tdf/sunflower/types/PropertyReader.java
package org.tdf.sunflower.types;

import lombok.Value;
import org.tdf.common.types.Uint256;
import org.tdf.common.util.HexBytes;
import org.tdf.sunflower.facade.PropertyLike;

import java.math.BigInteger;


@Value
public class PropertyReader {
    PropertyLike properties;

    public String getAsLowerCased(String property) {
        String s = properties.getProperty(property);
        return (s == null ? "" : s).trim().toLowerCase();
    }

    public String getAsNonNull(String property) {
        String s = properties.getProperty(property);
        return (s == null ? "" : s).trim();
    }

    public boolean getAsBool(String property) {
        return "true".equals(getAsLowerCased(property));
    }

    public int getAsInt(String property) {
        String s = properties.getProperty(property);
        if (s == null)
            throw new RuntimeException("read property " + property + " failed, property not found");
        return Integer.parseInt(s.trim());
    }

    public Uint256 getAsU256(String property, Uint256 defaultValue) {
        String s = properties.getProperty(property);
        if (s == null || s.trim().isEmpty()) {
            return Uint256.ZERO;
        }
        s = s.trim().toLowerCase();
        BigInteger b = s.startsWith("0x") ? new BigInteger(s.substring(2), 16) : new BigInteger(s);
        return Uint256.of(b);
    }

    public int getAsInt(String property, int defaultValue) {
        String s = properties.getProperty(property);
        if (s == null || s.trim().isEmpty())
            return defaultValue;
        return Integer.parseInt(s.trim());
    }

    public HexBytes getAsPrivate(String property) {
        String s = properties.getProperty(property);
        if (s == null)
            throw new RuntimeException("property " + property + " not found");
        HexBytes k = HexBytes.fromHex(s.trim());
        if (k.size() != 32)
            throw new RuntimeException("invalid private key size: " + k);
        return k;
    }

    public long getAsLong(String property) {
        String s = properties.getProperty(property);
        if (s == null || s.trim().isEmpty())
            throw new RuntimeException("read property " + property + " failed, property not found");
        return Long.parseLong(s.trim());
    }

    public HexBytes getAsAddress(String property) {
        String s = properties.getProperty(property);
        HexBytes minerCoinBase = HexBytes.fromHex(s);
        if (minerCoinBase.size() != Transaction.ADDRESS_LENGTH)
            throw new RuntimeException("invalid coinbase address " + minerCoinBase);
        return minerCoinBase;
    }
}

File: facade/src/main/java/org/tdf/sunflower/types/PropertyReader.kt
package org.tdf.sunflower.types

import org.tdf.common.types.Uint256
import org.tdf.common.util.Address
import org.tdf.common.util.HexBytes
import org.tdf.common.util.hex
import org.tdf.sunflower.facade.PropertyLike
import java.math.BigInteger

class PropertyReader(val properties: PropertyLike) {
    fun getAsLowerCased(property: String): String {
        val s = properties.getProperty(property)
        return (s ?: "").trim { it <= ' ' }.lowercase()
    }

    /**
     * get property as non null
     */
    fun getAsNonNull(property: String): String {
        val s = properties.getProperty(property)
        return (s ?: "").trim { it <= ' ' }
    }

    fun getAsBool(property: String): Boolean {
        return "true" == getAsLowerCased(property)
    }

    fun getAsInt(property: String): Int {
        val s = properties.getProperty(property)
            ?: throw RuntimeException("read property $property failed, property not found")
        return s.trim { it <= ' ' }.toInt()
    }

    fun getAsU256(property: String, defaultValue: Uint256): Uint256 {
        var s = properties.getProperty(property)
        if (s == null || s.trim { it <= ' ' }.isEmpty()) {
            return defaultValue
        }
        s = s.trim { it <= ' ' }.lowercase()
        val b = if (s.startsWith("0x")) BigInteger(s.substring(2), 16) else BigInteger(s)
        return Uint256.of(b)
    }

    fun getAsInt(property: String, defaultValue: Int): Int {
        val s = properties.getProperty(property)
        return s?.trim { it <= ' ' }?.takeIf { it.isNotEmpty() }?.toInt() ?: defaultValue;
    }

    fun getAsPrivate(property: String): HexBytes? {
        val s = properties.getProperty(property)
        val k = s?.trim { it <= ' ' }?.hex()
        if (k != null && k.size != 32) throw RuntimeException("invalid private key size: $k")
        return k
    }

    fun getAsLong(property: String, defaultValue: Long): Long {
        val s = properties.getProperty(property)
        if (s == null || s.trim { it <= ' ' }.isEmpty())
            return defaultValue
        return s.trim { it <= ' ' }.toLong()
    }

    fun getAsLong(property: String): Long {
        val s = properties.getProperty(property)
        if (s == null || s.trim { it <= ' ' }
                .isEmpty()) throw RuntimeException("read property $property failed, property not found")
        return s.trim { it <= ' ' }.toLong()
    }

    fun getAsAddress(property: String): Address? {
        val s = properties.getProperty(property)
        val addr = s?.lowercase()?.hex()
        if (addr != null && addr.size != Transaction.ADDRESS_LENGTH)
            throw RuntimeException("invalid address $addr length is ${addr.size}")
        return addr
    }

    fun getAsList(prefix: String): List<String> {
        val r = mutableListOf<String>()
        var i = 0
        while (true) {
            val v = properties.getProperty("$prefix.$i") ?: break
            r.add(v)
            i++
        }
        return r
    }
}